Save packets for which we don't have keys for later

martinthomson opened the pull request:
In order to test this, I had to add support for the server sending prior
to the handshake completing.  The nice thing is that we can use
availability of transport parameters to drive this.  If we have them,
then the server will be able to send.

There is a change in the HTTP code that bothers me a little.  There is a
new event being generated.  I don't understand why that event wasn't
being produced before.

Closes #714.

agrover commented:
> There is a change in the HTTP code that bothers me a little. There is a
new event being generated. I don't understand why that event wasn't
being produced before.

Which event?

martinthomson commented:
The event that I was seeing was that "StreamWritable" for stream 0 (the request stream).  For some reason, the code was not expecting that event previously.  But I couldn't see any way that it would not be generated.  I was hoping that @ddragana could explain why that wasn't being seen previously.

agrover commented:
Just to record here from our talk yesterday, I'm still seeing consistent 1-2 decryption failure messages with quic.rocks:4433, which I wasn't expecting and wanted to understand better.

martinthomson commented:
OK, I checked quic.rocks and everything looks fine.  Without the patch we drop two packets, with the patch we drop one.

The first packet we drop is the 1-RTT packet that is coalesced with a Handshake packet.  With this change we eventually process that packet.

The second packet we drop is an Initial packet, but we have dropped the Initial keys when that arrives, so we should be dropping that.

I've rebased this and tweaked the logic slightly.

martinthomson commented on neqo-transport/src/connection.rs:
```diff
@@ -5358,6 +5432,106 @@ mod tests {
         assert!(matches!(client.state(), State::Draining { .. }));
     }
 
+    /// Test that a server that coalesces 0.5 RTT with handshake packets
+    /// doesn't cause the client to drop application data.
```
```suggestion
    /// Test that a server can send 0.5 RTT application data.
```

agrover commented:
:+1: please merge once conflicts resolved.
